File: lessons/180-go-golang-vs-python-performance-benchmark-kubernetes-opentelemetry-prometheus-s3postgres/python-app/images.py
import time
import uuid
from datetime import datetime
from opentelemetry.trace import set_span_in_context


class Image:
    """Image represents the image uploaded by the user."""

    def __init__(self):
        """Create a default constructor for the Image."""

        self.uuid = self._generate_uuid()
        self.last_modified = self._get_time()

    def _generate_uuid(self):
        """Generate random UUID."""

        return str(uuid.uuid4())

    def _get_time(self):
        """Return current timestamp."""

        return datetime.now()


def download(s3, bucket, key, span, tracer, request_duration):
    """Downloads S3 image and returns last modified date."""

    # Get current time to record request duration.
    start = time.time()

    # Get current context to create child span id.
    context = set_span_in_context(span)

    # Create child span id for the S3 get request.
    with tracer.start_as_current_span("S3 GET", context) as span:

        # Download file from the S3 object store.
        object = s3.get_object(Bucket=bucket, Key=key)

        # Read all the image bytes returned by AWS.
        object['Body'].read()

        # Get current time to record request duration.
        end = time.time() - start
        request_duration.labels('s3').observe(end)

        return object['LastModified'], span


def save(image, table, pool, span, tracer, request_duration):
    """Inserts a newly generated image into the Postgres database."""

    # Get current time to record request duration.
    start = time.time()

    # Get current context to create child span id.
    context = set_span_in_context(span)

    # Create child span id for the DB Insert.
    with tracer.start_as_current_span("SQL INSERT", context) as span:

        # Connect to an existing database
        with pool.connection() as conn:

            # Open a cursor to perform database operations
            with conn.cursor() as cur:

                # Prepare the database query.
                query = f'INSERT INTO {table} (image_uuid, last_modified) VALUES (%s, %s)'

                # Send the query to PostgreSQL.
                cur.execute(query, (image.uuid, image.last_modified))

                # Make the changes to the database persistent
                conn.commit()

                # Get current time to record request duration.
                end = time.time() - start
                request_duration.labels('db').observe(end)

File: lessons/180-go-golang-vs-python-performance-benchmark-kubernetes-opentelemetry-prometheus-s3postgres/go-app/images.go
package main

import (
	"context"
	"fmt"
	"io"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/prometheus/client_golang/prometheus"
)

// Image represents the image uploaded by the user.
type Image struct {
	// ImageUUID is the unique ID of the image.
	ImageUUID string

	// LastModified is the timestamp when the image was last modified.
	LastModified time.Time
}

// NewImage creates a new image.
func NewImage() *Image {
	// Generate a new UUID for the image.
	id := uuid.New().String()

	// Simulate the last modified date.
	lastModified := time.Now()

	// Create an image with the generated ID and timestamp.
	image := &Image{
		ImageUUID:    id,
		LastModified: lastModified,
	}

	return image
}

// Save inserts a newly generated image into the Postgres database.
func Save(c *Image, table string, dbpool *pgxpool.Pool, m *metrics, ctx context.Context) error {
	// Create a new CHILD span to record and trace the request.
	ctx, span := tracer.Start(ctx, "SQL INSERT")
	defer span.End()

	// Get the current time to record the duration of the request.
	now := time.Now()

	// Prepare the database query to insert a record.
	query := fmt.Sprintf("INSERT INTO %s VALUES ($1, $2)", table)

	// Execute the query to create a new image record.
	_, err := dbpool.Exec(context.Background(), query, c.ImageUUID, c.LastModified)
	if err != nil {
		return fmt.Errorf("dbpool.Exec failed: %w", err)
	}

	// Record the duration of the insert query.
	m.duration.With(prometheus.Labels{"op": "db"}).Observe(time.Since(now).Seconds())

	return nil
}

// download downloads S3 image and returns last modified date.
func download(sess *session.Session, bucket string, key string, m *metrics, ctx context.Context) (*time.Time, context.Context, error) {
	// Create a new CHILD span to record and trace the request.
	ctx, span := tracer.Start(ctx, "S3 GET")
	defer span.End()

	// Get the current time to record the duration of the request.
	now := time.Now()

	// Create a new S3 session.
	svc := s3.New(sess)

	// Prepare the request for the S3 bucket.
	input := &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	}

	// Send the request to the S3 object store to download the image.
	output, err := svc.GetObject(input)
	if err != nil {
		return nil, nil, fmt.Errorf("svc.GetObject failed: %w", err)
	}

	// Read all the image bytes returned by AWS.
	_, err = io.ReadAll(output.Body)
	if err != nil {
		return nil, nil, fmt.Errorf("io.ReadAll failed: %w", err)
	}

	// Record the duration of the request to S3.
	m.duration.With(prometheus.Labels{"op": "s3"}).Observe(time.Since(now).Seconds())

	return output.LastModified, ctx, nil
}
